Kernel/Devices/AHCIDiskDevice.cpp

```diff
@@ -0,0 +1,120 @@
+* Copyright (c) 2020, Ben Jilks <benjyjilks@gmail.com>
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include <Kernel/Devices/AHCIChannel.h>
+#include <Kernel/Devices/AHCIDiskDevice.h>
+#include <Kernel/FileSystem/FileDescription.h>
+#include <Kernel/VM/MemoryManager.h>
+
+namespace Kernel {
+
+#define SECTOR_SIZE 512
+#define MAX_SECTORS 16
+#define BUFFER_SIZE MAX_SECTORS* SECTOR_SIZE
+
+#define LSB(x) ((x)&0xFF)
+#define MSB(x) (((x) >> 8) & 0xFF)
+#define MIN(a, b) (int)(a) < (int)(b) ? (a) : (b)
+
+AHCIDiskDevice::AHCIDiskDevice(AHCIPort* port, unsigned major, unsigned minor)
+    : BlockDevice(major, minor, 512)
+    , m_port(port)
+{
+    // Map a new region of continues memory for the buffer to go
+    m_buffer_region = MM.allocate_kernel_region(
+        PAGE_ROUND_UP(BUFFER_SIZE), "Data Buffer",
+        Region::Access::Read | Region::Access::Write);
+    m_buffer = m_buffer_region->vaddr();
+    m_buffer_physical = m_buffer_region->vmobject().physical_pages()[0]->paddr();
+
+#ifdef DEBUG_AHCI
+    kprintf("AHCIDiskDevice: Create buffer at Virtual Address %x and Physical Address %x\n",
+        m_buffer.get(), m_buffer_physical.get());
+#endif
+}
+
+bool AHCIDiskDevice::read_blocks(unsigned index, u16 count, u8* out)
+{
+    auto& slot = m_port->find_free_slot();
+    bool success = slot.issue_read(*m_port,
+        index, count, (u16*)m_buffer_physical.get());
+
+    memcpy(out, (void*)m_buffer.get(), count * 512);
+    return success;
+}
+
+bool AHCIDiskDevice::write_blocks(unsigned index, u16 count, const u8*)
+{
+    kprintf("AHCI: Write %i sectors to %i\n", count, index);
+    return true;
+}
+
+ssize_t AHCIDiskDevice::read(FileDescription& fd, u8* out, ssize_t len)
```

Comment:
Let's abstract the technical details from `AHCIDiskDevice`, so only the AHCI controller class will know how to deal with the hardware. Also, we don't want to repeat the `PATADiskDevice`, so please make a class called `StorageDevice`, so `PATADiskDevice` will inherit from it, and `AHCIDiskDevice` will do that too.